consensus/service/src/validators.rs

```diff
@@ -101,47 +100,54 @@ impl<L: Ledger> TxManagerUntrustedInterfaces for DefaultTxManagerUntrustedInterf
         Ok(())
     }
 
-    /// Combines a set of "candidate values" into a "composite value". This assumes all values are well
-    /// formed and safe to append to the ledger individually.
-    /// ASSUMES VALUES ARE SORTED.
+    /// Combines a set of "candidate values" into a "composite value".
+    ///
+    /// This assumes each value is well-formed and safe to append to the ledger individually.
     ///
     /// # Arguments
     /// * `tx_contexts` - "Candidate" transactions. Each is assumed to be individually valid.
     /// * `max_elements` - Maximum number of elements to return.
     ///
     /// Returns a bounded, deterministically-ordered list of transactions that are safe to append to the ledger.
-    fn combine(
-        &self,
-        tx_contexts: &[&WellFormedTxContext],
-        max_elements: usize,
-    ) -> BTreeSet<TxHash> {
-        // Allow transactions that do not introduce key image double-spends.
-        let mut allowed_hashes = BTreeSet::new();
-        let mut used_key_images = HashSet::default();
-
-        for tx_context in tx_contexts.iter() {
-            if allowed_hashes.len() >= max_elements {
-                // Enforce maximum size.
-                break;
-            }
+    fn combine(&self, tx_contexts: &[&WellFormedTxContext], max_elements: usize) -> Vec<TxHash> {
+        // Dedup by TxHash
+        let mut candidates: Vec<&WellFormedTxContext> = {
+            let hash_to_candidate: HashMap<&TxHash, &WellFormedTxContext> = tx_contexts
```

Comment:
Why is this building a HashMap and then discarding the keys and just grabbing the values?

Reply:
The intent is to remove any duplicates contained in the collection `tx_contexts`. Building a HashMap and then taking the values has roughly the same "dedup-ing" effect as building a HashSet and then taking its values. I think I took the HashMap route instead of the HashSet route because a HashSet would allow two instances of WellFormedTxContext that contain the same TxHash but have other differing fields, while the HashMap keyed by TxHash would not. 

In retrospect, that's not really an important difference. I'll switch it to a HashSet, which should be more readable.